statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/StatisticsService.java

```diff
@@ -38,248 +34,270 @@
-  }
+    /**
+     * It queries the requested data taking into account the filters
+     *
+     * @param statisticsRequest The wrapper that contains the datasetId and the filters and its respective values to query
+     * @return An object containing the result of the filtering and the available options based on the filtering performed
+     */
+    public FilteringResult queryDataWithFilters(FiltersWrapper statisticsRequest)
+            throws BreakdownDeclarationFailException {
+        //Prepares the query with the requested filters
+        StatisticsQuery readyQuery = prepareFilteringQuery(statisticsRequest);
+
+        //Executes the query and obtains the result
+        StatisticsData queryResult = readyQuery.queryForStatistics();
+
+        // The available options are based on the query that was executed previously
+        Map<MongoStatisticsField, Set<String>> valueAvailableOptions = prepareValueFilteringOptionsQuery(
+                readyQuery);
+        Map<MongoStatisticsField, Optional<ValueRange>> rangeAvailableOptions = prepareRangeFilteringOptionsQuery(
+                readyQuery);
+
+        int totalRecords = queryResult.getRecordCount();
+
+        StatisticsResult statisticsAllRecordsResult = new StatisticsResult(STATISTICS_RESULT_ROOT_VALUE,
+                totalRecords, calculatePercentage(totalRecords, totalRecords));
+
+        // If there are breakdowns present, parse them and set them in the final result
+        if (!queryResult.isBreakdownListEmpty()) {
+            Pair<FacetValue, List<StatisticsResult>> parsedBreakdown = parseBreakdownsAsOutput(
+                    queryResult, totalRecords);
+            statisticsAllRecordsResult.setBreakdowns(
+                    new BreakdownResult(parsedBreakdown.getKey(), parsedBreakdown.getValue()));
+        }
+
+        // Converting the result of available options to object FilteringOptions
+        FilteringOptions filteringOptions = new FilteringOptions();
+        valueAvailableOptions.forEach(
+                (key, value) -> key.getValueFilterSetter().accept(filteringOptions, value));
+        rangeAvailableOptions.forEach((key, value) -> key.getRangeFilterSetter()
+                .accept(filteringOptions,
+                        new StatisticsRangeFilter(value.map(ValueRange::getFrom).orElse(null),
+                                value.map(ValueRange::getTo).orElse(null))));
+
+        return new FilteringResult(statisticsAllRecordsResult, filteringOptions);
+    }
 
-  private double calculatePercentage(double totalCount, double count) {
-    return totalCount <= 0 ? 0 : Double.parseDouble(PERCENTAGE_FORMAT.format((count / totalCount) * 100.0));
-  }
+    /**
+     * Gathers all different urls associated to a given rights category
+     *
+     * @param category The right category to query
+     * @return A set of different rights urls associated to a given category
+     */
+    public Set<String> getRightsUrlsWithCategory(RightsCategory category) {
```

Comment:
I thought of this scenario as well. This was my train of thought at the time: Since the database is updated weekly, then the service will also restart, the Map will be a in clean state and capture the new values. But I just realized that is not true! If we want to keep this logic somehow, then these are some possibilities to consider (that I can think of):

1. We keep the code as it is and we will need to restart the service as well every time we update the database, which doesn't make sense to do that 
2. We can create a listener so that every time the script is run, we drop all saved values in the map so it's in a clean state and can capture the new values (if there are any), but not sure if this affects performance in any way
3. We can scratch all of this, trust mongodb's performance and just do a query every time this method is called.

I'm inclined to go with the 3rd option, but what do you think?

Reply:
I think there is another option as well.
Create the map for all rights statements at the application startup and also have a re-occurring thread that will do that again every half an hour or so. 
So the map will always be full and every 30mins we have a new state of it.